open-craft/edx-platform
# Move celery check task to the high priority queue

Currently, LMS uses 3 Celery workers: lms_default_1, lms_high_1 and
lms_high_mem_1. Each Celery worker sends messages to a single queue:
edx.core.default, edx.core.high and edx.core.high_mem, respectively.
The number of child processes per Celery worker is set to 1. Due to
this configuration, any task in a queue blocks all other tasks.

Currently, the Celery check task submitted by the /heartbeat?extended
LMS HTTP API endpoint runs in the default queue. When some slow task
(eg course grades creation) is sent to the default queue, it will
block the Celery check task, which will expire and the heartbeat endpoint
will fail. This patch moves the task to another queue which has
only shorter tasks and in which this problem will not occur.

Test instructions:
- Create an ironwood.2 VM in OCIM with this edx-platform branch. I recommend using production OCIM because it uses a exclusive rabbitmq for each VM, while staging uses one shared by all VMs. Besides that, VM creation is failing a lot in staging OCIM in the last few days, probably due to OVH issues.
- Log in the VM using ssh: ssh <user>@<vm_ip_address>
- Run "sudo su"
- Add an infinite Celery task to /edx/app/edx-platform/openedx/core/djangoapps/heartbeat/tasks.py:

```
import time
@task()
def infinite_task():
    while True:
        time.sleep(3)
    return True
```

- Call the new task in a view by adding the following to openedx/core/djangoapps/heartbeat/views.py

```
from .tasks import infinite_task
def infinite_task_view(request):
    infinite_task.apply_async()
    return JsonResponse()
```

- Create a new HTTP API endpoint which calls the new view by add the following to the urlpatterns list in /edx/app/edx-platform/openedx/core/djangoapps/heartbeat/urls.py:

```
url(r'^/infinite_task$', infinite_task_view, name='infinite_task')
```

- Restart the Celery workers used by LMS: /edx/bin/supervisorctl restart {edxapp_worker:lms_default_1,edxapp_worker:lms_high_1,edxapp_worker:lms_high_mem_1}  

 - Restart LMS: /edx/bin/supervisorctl restart lms

- Navigate to <openedx_instance_base_url>/heartbeat/infinite_task in a web browser
- Make sure the infinite Celery task is running in the LMS default queue by executing: /edx/bin/{python,manage}.edxapp lms --settings=production celery inspect active

Sample output:
-> celery@edx.lms.core.default.edxapp-alanoe-bb2223-appserver-1: OK
    * {u'args': u'[]', u'time_start': 507476.889938013, u'name': u'openedx.core.djangoapps.heartbeat.tasks.infinite_task', u'delivery_info': {u'priority': 0, u'redelivered': False, u'routing_key': u'edx.lms.core.default', u'exchange': u'edx.lms.core'}, u'hostname': u'celery@edx.lms.core.default.edxapp-alanoe-bb2223-appserver-1', u'acknowledged': True, u'kwargs': u'{}', u'id': u'ee4d395a-f301-4dd8-afc0-89a3dbc9cf46', u'worker_pid': 26477}
-> celery@edx.cms.core.default.edxapp-alanoe-bb2223-appserver-1: OK
    - empty -
-> celery@edx.lms.core.high_mem.edxapp-alanoe-bb2223-appserver-1: OK
    - empty -
-> celery@edx.lms.core.high.edxapp-alanoe-bb2223-appserver-1: OK
    - empty -
-> celery@edx.cms.core.high.edxapp-alanoe-bb2223-appserver-1: OK
    - empty -


 - Navigate to <openedx_instance_base_url>/heartbeat?extended in a web browser. Make sure the HTTP API request returns a HTTP response with status code 200 and a JSON output with the "status" attribute in the "celery" attribute set to "true". Sample JSON output:

{
  "modulestore": {
    "status": true, 
    "message": "OK"
  }, 
  "celery": {
    "status": true, 
    "message": "{'time': '0.325976133347'}"
  }, 
  "forum": {
    "status": true, 
    "message": "OK"
  }, 
  "sql": {
    "status": true, 
    "message": "OK"
  }
}
